File: lib/core/security/review_write.dart
import '../models/enums.dart';

/// Fail-closed review writes. Matches live Dahr LY:
///
/// * **Consumer** — INSERT only. Payload must not set `is_hidden`.
/// * **Admin** — hide via `{is_hidden: true}` only. Flutter has no hide UI.
abstract final class ReviewWrite {
  static const String hiddenColumn = 'is_hidden';

  static bool canInsert(UserRole actor) =>
      actor == UserRole.consumer || actor == UserRole.admin;

  /// Consumers never UPDATE reviews (`is_hidden` is admin-only).
  static bool canUpdate(UserRole actor) => actor == UserRole.admin;

  static bool touchesHidden(Map<String, dynamic> payload) =>
      payload.containsKey(hiddenColumn);

  /// Leave-review insert body. Throws if [hiddenColumn] is present.
  static Map<String, dynamic> consumerInsert(Map<String, dynamic> payload) {
    if (touchesHidden(payload)) {
      throw StateError('write_rejected');
    }
    return payload;
  }

  /// Any consumer UPDATE (including hide) is rejected.
  static Map<String, dynamic> consumerUpdate([
    Map<String, dynamic>? _,
  ]) {
    throw StateError('write_rejected');
  }

  /// Admin hide patch — the only review UPDATE the clients may send.
  static Map<String, dynamic> adminHidePatch() => {hiddenColumn: true};
}

File: admin/src/lib/review-write.ts
/** Fail-closed review writes. Mirrors Flutter `ReviewWrite`. Live Dahr LY:
 * reviews are insert-only; `is_hidden` is admin-only.
 */

export const REVIEW_HIDDEN_COLUMN = "is_hidden";

export function touchesHidden(
  payload: Record<string, unknown>,
): boolean {
  return Object.prototype.hasOwnProperty.call(payload, REVIEW_HIDDEN_COLUMN);
}

/** Leave-review insert body. Consumers must not send `is_hidden`. */
export function consumerReviewInsert<T extends Record<string, unknown>>(
  payload: T,
): T {
  if (touchesHidden(payload)) {
    throw new Error("write_rejected");
  }
  return payload;
}

/** Consumers never UPDATE reviews, including hide. */
export function consumerReviewUpdate(): never {
  throw new Error("write_rejected");
}

/** Admin hide patch — the only review UPDATE this dashboard may send. */
export function adminHideReviewPatch(): { is_hidden: true } {
  return { is_hidden: true };
}
